### DISCUSSION TITLE
Error on Running Example

### DISCUSSION THREAD

**Discussion by kaypersb** [alexandercerutti/passkit-generator#94] [Q&A] [answered]
Asking for help, newbie here :)

I tried to run an example. I downloaded my certificates (wwdr and Certificates.p12) but running into an error when I launch http://localhost:8080/pkpasses/exampleBooking.pass. Getting this error: Could not add PKCS#7 signer; no private key specified. Can somebody help? I was able to manually create a pass using this certificates but not sure why I can't using the example provided. Thanks!

Steps I did:
1. DL Certificates.p12 thru keychain (entered a password when exporting)
2. DL WWDR from apple site (used G1 - no password prompted when I exported to .pem)
3. Created signerCert and signerKey files using the openssl scripts provided.
4. Go to http://localhost:8080/pkpasses/exampleBooking.pass - error encountered. 

Thank you.

**Comment by alexandercerutti** [comment, marked as answer]
Hey there @kaypersb, did you change where the certificates are read from or did you change anything in the example?

**Reply by kaypersb** [reply]
Hi @alexandercerutti, I created a folder in the root - passkit-generator. I initially encountered this error so I thought I created the folder in the wrong directory. 

```sh
(node:30774) UnhandledPromiseRejectionWarning: Error: ENOENT: no such file or directory, open '/Users/kaye/Documents/GitHub/passkit-generator/certificates/signerCert.pem'
```

After I created this folder and included the certificates, the error "Could not add PKCS#7 signer; no private key specified" showed up.
I did not change anything from the examples folder. Please let me know if I installed it incorrectly - I ran npm install at self-hosted folder. Thank you, appreciate your response!

Error stack:

```sh
Webserver started.
Error: Could not add PKCS#7 signer; no private key specified.
    at Object.addSigner (/usr/local/lib/node_modules/passkit-generator/node_modules/node-forge/lib/pkcs7.js:262:15)
    at Object.create (/usr/local/lib/node_modules/passkit-generator/lib/Signature.js:40:15)
    at PKPass.[pass.close] (/usr/local/lib/node_modules/passkit-generator/lib/PKPass.js:436:43)
    at PKPass.getAsBuffer (/usr/local/lib/node_modules/passkit-generator/lib/PKPass.js:451:42)
    at /usr/local/lib/node_modules/passkit-generator/lib/PKPass.js:97:51
    at Array.map (<anonymous>)
    at Function.pack (/usr/local/lib/node_modules/passkit-generator/lib/PKPass.js:97:32)
    at /Users/kaye/Documents/GitHub/passkit-generator/examples/self-hosted/src/PKPasses.ts:132:27
```

**Reply by alexandercerutti** [reply]
That's kinda weird. So you can confirm that inside `passkit-generator/certificates` there are three files: `signerKey.pem`, `signerCert.pem` and `wwdr.pem`? Can you confirm that `signerKey.pem` has content (and it is valid)? That's the only reason that is coming to my mind for that reason to happen.

**Reply by kaypersb** [reply]
Here's my certificates files:

```sh
(base) Kayes-MacBook-Pro:certificates kaye$ ls -l
total 40
-rw-r--r--@ 1 kaye  staff  3267 Dec  7 01:31 Certificates.p12
-rw-r--r--@ 1 kaye  staff  1493 Dec  7 00:52 WWDR.pem
-rw-------  1 kaye  staff  2545 Dec  7 00:51 signerCert.pem
-rw-------  1 kaye  staff  2012 Dec  7 00:51 signerKey.pem
```

I followed all steps provided in your wiki. Also, can you check if my folder structure is correct? I installed it in self-hosted folder so here it is:

```sh
(base) Kayes-MacBook-Pro:self-hosted kaye$ ls -l
total 136
-rw-r--r--   1 kaye  staff   2545 Dec  7 00:08 README.md
drwxr-xr-x   4 kaye  staff    128 Dec  7 01:23 models
drwxr-xr-x  71 kaye  staff   2272 Dec  7 00:11 node_modules
-rw-r--r--   1 kaye  staff  55533 Dec  7 00:11 package-lock.json
-rw-r--r--   1 kaye  staff    831 Dec  7 00:08 package.json
drwxr-xr-x  12 kaye  staff    384 Dec  7 00:08 src
-rw-r--r--   1 kaye  staff    211 Dec  7 00:08 tsconfig.json

(base) Kayes-MacBook-Pro:self-hosted kaye$ pwd
/Users/kaye/Documents/GitHub/passkit-generator/examples/self-hosted
```

**Reply by kaypersb** [reply]
Hi! @alexandercerutti,
I got passed thru the error - I changed my passin code when generating the signerKey.pem to 123456. I read it in the shared.ts file, it's hardcoded to 123456. Yay :) 

My next issue is - when I try to go to this url `localhost:8080/pkpasses/examplePass.pass`, it downloads the file examplePass.pass_20211209.pkpasses, when I unzip them, I can't open the actual pass. I tried to unzip these files (renamed them to .zip) just to see what's inside. I can see all the files (logo, pass.json), but can't open them as a pass.

```sh
(base) Kayes-MacBook-Pro:examplePass.pass_20211209 kaye$ ls -l
total 96
-rw-rw-r--@ 1 kaye  staff  10854 Nov 30  1979 packed-pass-1.pkpass
-rw-rw-r--@ 1 kaye  staff  10854 Nov 30  1979 packed-pass-2.pkpass
-rw-rw-r--@ 1 kaye  staff  10848 Nov 30  1979 packed-pass-3.pkpass
-rw-rw-r--@ 1 kaye  staff  10855 Nov 30  1979 packed-pass-4.pkpass
```

I tried this `http://localhost:8080/localize/exampleBooking.pass` - but got error: `Cannot proceed creating the pass because transitType is missing on your boardingPass`.

Let me know if I am missing anything! thanks :)

**Reply by alexandercerutti** [reply]
Hey @kaypersb, asking you to look at the signerKey passphrase would have been my next move. :) Glad you solved this issue.

How are you attempting to open them? I mean, yes, you are generating a bundle of passes (a.k.a. PKPasses). Please note that `.pkpasses` bundles can be actually opened only from Safari as per iOS 15.1 (or, maybe, an app that supports them, but I don't know if there's an iOS API for this).

Opening a `.pkpass` should be supported in any application. Which application are you using to open the single `.pkpass`?

Anyway, look at [Wiki guide for troubleshooting](https://github.com/alexandercerutti/passkit-generator/wiki/Troubleshooting-(Self-help)) so you can learn how to see what's wrong with passes, by using a mac.

For the `localize` example, yes, the model you are loading is missing the transitType property, so of course it is not going to work. I might think to add a condition to add a default transitType if needed. Try using the other model provided.

**Reply by kaypersb** [reply]
Thanks @alexandercerutti for your patience :) I am still trying to figure it out but still stuck :( When I go to the url http://localhost:8080/localize/exampleBooking.pass (on mac), the files are downloaded automatically. If I open, it asks for an app to open with. It doesn't allow safari tho. I tried to change them into zip files, then open each .pkpass file and gives me this error : The pass “packed-pass-1.pkpass” could not be opened. When I created a pass manually (using my own .json files), I can open .pkpass without error. 

When I check the console.app this is the error which I don't understand :( This is the same error I get when I try for model examplePass.pass. Should I be opening it using my iphone? I tried to send it via airdrop but still does not open. 

Dec  9 22:26:42 Kayes-MacBook-Pro com.apple.xpc.launchd[1]: Coalition Cache Hit: app<application.com.apple.Pass
-Viewer.1152921500312180543.1152921500312180552(502)> [3722]
Dec  9 22:26:50 Kayes-MacBook-Pro com.apple.xpc.launchd[1] (com.apple.mdworker.shared.03000000-0400-0000-0000-000000000000[34263]): Service exited due to SIGKILL | sent by mds[99]
Dec  9 22:26:50 Kayes-MacBook-Pro com.apple.xpc.launchd[1] (com.apple.mdworker.shared.07000000-0100-0000-0000-000000000000[34264]): Service exited due to SIGKILL | sent by mds[99]
Dec  9 22:26:52 Kayes-MacBook-Pro Google Chrome Helper[977]: Libnotify: notify_register_coalesced_registration failed with code 9 on line 2835
Dec  9 22:27:46 Kayes-MacBook-Pro com.apple.xpc.launchd[1]: Coalition Cache Hit: app<application.com.apple.Pass-Viewer.1152921500312180543.1152921500312180552(502)> [3722]
Dec  9 22:27:52 Kayes-MacBook-Pro Google Chrome Helper[977]: Libnotify: notify_register_coalesced_registration failed with code 9 on line 2835

**Reply by alexandercerutti** [reply]
Uh, what the hell is that error?!

Anyway I think everything is going to get a lot confused. Do you mind separating things, in a visible way with text, you are doing on iOS from things you are doing on Mac and specify by row the example you are running, please? 😵‍💫

**Reply by kaypersb** [reply]
Ok maybe let's do a step by step? First thing I do is run this script
(base) Kayes-MacBook-Pro:self-hosted kaye$ DEBUG=* npm run example

Then open the url on safari using my mac - localhost:8080/pkpasses/exampleBooking.pass
I get this log in my terminal:

```
  express:router dispatching GET /pkpassfrom/examplebooking.pass +13s
  express:router query  : /pkpassfrom/examplebooking.pass +1ms
  express:router expressInit  : /pkpassfrom/examplebooking.pass +0ms
  express:router jsonParser  : /pkpassfrom/examplebooking.pass +2ms
  body-parser:json skip empty body +0ms
```

There's nothing downloaded. And nothing is coming out of the safari.

When I open this url in safari : localhost:8080/pkpassfrom/examplebooking.pass
It automatically downloads the .pkpasses file. When I try to open - it asks for an app to open it with. I rename it to .zip files so I can open the contents. I see there are multiple .pkpass files. When I double click the .pkpass file - it shows the error The pass “packed-pass-1.pkpass” could not be opened.

Hope it's clearer now. Let me know if you need more information. All of this is being opened using the mac. Didn't use my mobile to open the files. Thank you so much!

**Reply by alexandercerutti** [reply]
It's clearer but...

That's weird! I cannot explain such a thing, but it seems an issue with Safari 🤔
Does you Safari require manual confirmation for download? Or maybe does it prompts on other windows, the modal to select where to save a file?

After extracting, try to open `Console.app` and filter for "Passkit" or "Pass Viewer" and then try to open the .pkpass. Report the logs here.

Anyway, `DEBUG=*` (as you can read) cannot be used anymore to debug passkit-generator. It was available in v2.

**Reply by kaypersb** [reply]
Hi @alexandercerutti... I did a brute froce to manually check what is wrong with my pkpass. What I did was to update the pass.json files inside one of the pkpasses. I updated it to my own passtype/team identifiers, created the signature again (using my own certificates) and recreated .pkpass file and finally got to open it! I was thinking since it was my own certificates that are being installed - then the credentials inside the pass.json should be mine too... Should I be doing something in the model files first before I ran it? I was under the impression that I shouldn't be changing anything and just run it... Let me know if there is something I missed or I might have misunderstood something in the process.

**Reply by alexandercerutti** [reply]
Oh, my! I should have specified this in the examples readme, but I thought it was obvious. Passkit-generator requires a read of Apple documentation before using it.

Of course, teamIdentifier and passTypeIdentifier are linked to your certificates, because, if you analyze the signerCert through OpenSSL, you can see it contains these two information. So if they do not match, Wallet and macOS pass viewer think the pass might have been altered, while one of the principles of Passkit is integrity.

Did you create the signature again manually? How?

Great you have solved! I must understand which is the best way to write in examples what you did discover.

P.s. if you open Pkpasses archive through Safari on iPhone, do you see the passes bugged too, like in the screenshots I've attached in the example code header comment? Let me know.

**Reply by kaypersb** [reply]
Yes, I read the apple docs though I must admit - I didn't think this through and thought it would run as is! My bad. But yeah maybe you can add it in your readme for newbies like me! :) But super thankful for all your help!!! I really learned a lot! :)

I created the signature using this code:

```sh
$ openssl smime -binary -sign -certfile WWDR.pem -signer signerCert.pem -inkey signerKey.pem -in manifest.json -out signature -outform DER -passin pass:<your own pass>
```

I ran `openssl sha1` for the pass.json again to update the manifest.json.

Sorry alex, where should I see the screenshots in the pkpasses? I still cannot open them in safari in my iphone :(

**Reply by alexandercerutti** [reply]
Yeah, I'll add more details. Also, thank you for the command, I'll add a paragraph to the certificates guide!

Well, you have to ensure MacOS's firewall allows incoming connections for Terminal.app (I guess?), then start the server in your mac. Pick your IP address and paste it, along with the examples path, in your iOS 15 Safari. You should be able to get passes there.

The screenshots are available in the code here: https://github.com/alexandercerutti/passkit-generator/blob/master/examples/self-hosted/src/PKPasses.ts#L15

Let me know if you see them like I do (screenshots).

**Reply by alexandercerutti** [reply]
Hey, @kaypersb I've released a new version with a few changes and the improvements we talked about!
I've also added a whole new paragraph to the wiki: https://github.com/alexandercerutti/passkit-generator/wiki/Generating-Certificates#do-i-really-need-passkit-generator-to-create-apple-wallet-passes

Thanks for your (involuntary) contribution eh eh 😁

**Reply by kaypersb** [reply]
I can't see screesnhots here: 
https://github.com/alexandercerutti/passkit-generator/blob/master/examples/self-hosted/src/PKPasses.ts#L15

But I was already able to see my pkpass in my iphone. Yey! 

And you're welcome (hehe). Happy to have contributed unexpectingly ;)

**Reply by alexandercerutti** [reply]
@kaypersb there are some links inlined. Check better in the top comment 😄

Great!

**Reply by kaypersb** [reply]
Got it! Thanks again!!! :)
